keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This enhancement proposal covers implementing the structured configuration for the OIDC authenticator.
+OIDC authentication is essential part of Kubernetes, yet it has limitations in its current state. 
+Bellow we will discuss that limitation and propose solutions.
+
+# Motivation
+
+Structured config for OIDC authentication: noted in various contexts over the past few years. We want to migrate
+away from a flag-based config that is growing without bounds to a proper versioned config format. This would allow us to
+better support various features that have been requested.
+
+### Goals
+
+There are features users want to tune. We need to provide customization of the following:
+
+- Claims validation rules: current OIDC provider supports only audience claim validation and only by exact values
+- Claim mappings: it is only possible to pick a single value from a single claim and prefix groups
+- Use more than one OIDC provider: the only option, for now, is to use [Dex](https://dexidp.io/) (or similar software)
+  as a lightweight OIDC proxy to connect many providers to Kubernetes
+- Change authenticator settings without restarting kube-apiserver
+
+### Non-Goals
+
+- Monitoring
+
+## Proposal
+
+1. Add new `authentication` API object to parse a structured config file `OIDCConfiguration`.
+2. Add a single flag for kube-apiserver to point to the structured config file, automatically reload it on changes.
+3. Use an expression language to let users write their own logic for mappings and validation rules 
+  (expressions should be simple for common cases, yet powerful to cover most user stories).
+
+### Risks and Mitigations
+
+Since this is a new optional feature, no migration is required. Before the Stable release of the feature,
+we should provide examples of migrating from a flag-based config to a new structured config.
```

Comment:
a risk is destabilization or drift between the common aspects of the current flag-based OIDC impl and the config-file based implementation

a mitigation for that is to have the internal OIDC config / representation be able to be mapped to from both the flag-based and file-based configs, with the flag-based config stubbing out or no-oping the aspects that are not configurable via that mechanism

Reply:
We need to improve the test coverage around OIDC before we start trying to change the code that is used by the stable API IMO.  We also risk regressions in the stable API if we try to make it use CEL under the hood.

Reply:
> We need to improve the test coverage around OIDC before we start trying to change the code that is used by the stable API IMO

I'm on board with that

> We also risk regressions in the stable API if we try to make it use CEL under the hood.

I also agree we should not switch the implementation of existing aspects to CEL.

That said, it is important that the authenticator driven by this config be able to exactly match the behavior of the authenticator driven by CLI flags, so that people can migrate CLI config to file config as a first step with zero change in behavior.

The most straightforward way to ensure the config file can be used to drive identical behavior to the CLI flags is to:
1. have the config file be as expressive as the CLI flags without relying on CEL
2. have the fancy mapping/filtering/extraction aspects proposed here be optional/additional (and mutually exclusive with some of the static CLI options where it makes sense)
3. evolve the existing authenticator's code paths to add support for the new options in ways that are no-ops when those options are not specified